[metabase-lib] Port number formatting to CLJC; use it from JS

bshepherdson opened the pull request:
This adds a cross-platform number formatting library in CLJC, which has
(nearly) identical output in both JVM and JS environments.

The only known differences are:
- JVM Clojure has `BigDecimal` support, so isn't limited to a `double`'s
  range.
- Exact spelling/capitalization of the full names of currencies differs.
    - JVM: "7.34 US Dollar"
    - JS: "7.34 US dollars"

A couple of quirks I think we can improve in the future:
- Compact currency mode shows negatives as `$-xM`; regular currency as `-$12.34`. Do we care? (This is not a change.)
- Rounding is currently half-up in both languages (ie. `x.5` rounds to `x+1` in all cases). Java supports lots of options, but JavaScript does not. The proposed `roundingMode` option is exactly this but it's [not implemented anywhere](https://caniuse.com/mdn-javascript_builtins_intl_numberformat_numberformat_options_roundingmode_parameter) right now.
- See the [comment](https://github.com/metabase/metabase/pull/27114/files#diff-37b6692d818e9f6fc25bfbb8c9947fcf995452b50a189a870336218eaa7e3121R49) about currency formatting and forcing `"en"` locale in `internal/numbers.cljs`.
All of these are captured as TODOs in the code.

bshepherdson committed:
[metabase-lib] Port number formatting to CLJC; use it from JS

This adds a cross-platform number formatting library in CLJC, which has
(nearly) identical output in both JVM and JS environments.

The only known differences are:
- JVM Clojure has `BigDecimal` support, so isn't limited to a `double`'s
  range.
- Exact spelling/capitalization of the full names of currencies differs.
    - JVM: "7.34 US Dollar"
    - JS: "7.34 US dollars"

Some future work is required to fully land this. In particular, the
`currency_in_header` logic should be moved to the table formatters, not
implemented on the JS side of this change.

bshepherdson commented:
Feel free to add more FE folks; I'm not sure who to ask to review these changes.

bshepherdson commented on .github/workflows/frontend.yml:
```diff
@@ -61,6 +61,8 @@ jobs:
     - name: Prepare front-end environment
       uses: ./.github/actions/prepare-frontend
+    - run: yarn build-quick:cljs
+      name: Compile CLJS
```
This bit has been spun out into a separate PR, since merged.

camsaul commented:
I'm of the opinion that if we actually have a real use case for rounding in a different way than the normal mode we can cross that bridge when we get there

camsaul commented on shared/src/metabase/shared/formatting/internal/numbers.clj:
```diff
@@ -0,0 +1,109 @@
+    ;; (Intl.NumberFormat has an option `roundingMode` but it's new and not supported anywhere as of EOY2022.)
+    ;; Since Java is flexible, we match the HALF_UP behavior here.
+    (.setRoundingMode nf RoundingMode/HALF_UP)
+    (when (:minimum-fraction-digits options)
+      (.setMinimumFractionDigits nf (:minimum-fraction-digits options)))
+    (when currency
+      (.setCurrency nf (if (bad-currencies currency)
+                         ;; For the currencies the JVM doesn't support, we use USD and replace the symbols later.
+                         (Currency/getInstance "USD")
+                         (Currency/getInstance (name currency)))))
+    (when (:number-separators options)
+      (let [^DecimalFormat df nf
+            syms              (doto (.getDecimalFormatSymbols df)
+                                (.setDecimalSeparator  (first  (:number-separators options))))]
+        (if-let [grouping (second (:number-separators options))]
+          (.setGroupingSeparator syms grouping)
+          (.setGroupingUsed df false))
+        (.setDecimalFormatSymbols df syms)))
+    (reify
+      core/NumberFormatter
+      (format-number-basic [_ number]
+        (cond-> (.format nf (pre (bigdec (double number))))
+          ;; If running a "bad" currency Java doesn't support, replace the default symbol with the real one.
+          (and currency (bad-currencies currency))
+          (attach-currency-symbol nf locale currency)
+          ;; Handle the :currency-style option, which isn't supported natively on Java.
+          (and currency (not= (:currency-style options) "symbol"))
+          (apply-currency-style (.getCurrency nf) locale (:currency-style options) currency)))
+      (wrap-currency [_ text]
+        (let [sym (case currency
+                    :BTC "₿"
+                    (-> (name currency)
+                        (Currency/getInstance)
+                        (.getSymbol locale)))]
+          (str sym text)))
+      (split-exponent [_ formatted]
+        (let [^DecimalFormat df nf ;; Scientific mode always uses the DecimalFormat subclass.
+              exp (str/last-index-of formatted (.getExponentSeparator (.getDecimalFormatSymbols df)))]
+          {:mantissa (subs formatted 0 exp)
+           :exponent (subs formatted (inc exp))})))))
```
I don't really love 70-line functions, they make my eyes bleed. It would be nice if it were possible to break out some of the different parts here into different subroutines. At the very least maybe you could do something like 

```clj
```clj
(defn- set-currency! [^NumberFormatter fomatter currency]
  ...)

(doto (number-formatter-for-style (:style options))
  ...
  (set-currency! (:currency options))
  ...)
```

bshepherdson replied:
I've broken these up in both languages. They're still chunky, but now it's a few one-line `let` bindings and then the three methods under `reify`, which is hard to reduce further.

camsaul reviewed (approved):
The backend stuff looks good to me, the only suggestion I'd make is to break up the mega function that creates the formatter in the Clojure implementation into a few smaller functions to make it a little easier to read. 

Let's get this merged 💪 🚀

ranquild commented on frontend/src/metabase/lib/formatting/numbers.tsx:
```diff
@@ -1,253 +1,51 @@
-  );
-  if (options.jsx) {
-    const [m, n] = exp.split("e");
+  let formatted = format_number(number, kebab_keys(options));
```
Doesn't seem right that we need to call `kebab_keys` before passing something to a cljs function. Could this be handled internally?

bshepherdson committed:
Review comments

- broke up the megafunctions in internal/numbers (both languages)
- more general options map normalization, handled on CLJC side.
  - Broad goal is that JS/TS wrapper functions should be trivial
    wrappers plus JS-specific logic.
  - Any shareable logic, like massaging the inputs, should live in the CLJS.

bshepherdson committed:
better standard aliases

metamben commented on shared/src/metabase/shared/util.cljc:
```diff
@@ -12,3 +15,35 @@
+  "Given a string or keyword, in snake_case or kebab-case, normalize it to a :kebab-case keyword.
+
+  Note that namespaces on input keywords are dropped! This is intended for turning idiomatic JSON
+  `\"snake_case_string_keys\"` into idiomatic Clojure `:kebab-case-keywords`."
```
Would't it lead to surprises if it handled namespaces?

bshepherdson replied:
Are you suggesting that I tweak it to preserve the namespaces of any incoming namespaced keys, as a Least Surprise thing? That's a bit more complex logic but I agree it's probably a good idea.

bshepherdson replied:
Done, take a look.

metamben commented on shared/src/metabase/shared/util/currency.cljc:
```diff
@@ -873,9 +873,19 @@
           :name_plural "Zambian kwachas"}]])
 
+(def currency-map
+  "The currencies as a Clojure map. Wrapped in [[delay]] so it's only computed on demand."
```
Why is it important to compute it only on demand?

bshepherdson replied:
I dunno, but it's easy to use `delay` to preserve the caching logic in the [original](https://github.com/metabase/metabase/blob/f5d85e8131903976d4b4194f7f5de728b522fe9c/frontend/src/metabase/lib/formatting/currency.js#L8).

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers.clj:
```diff
@@ -0,0 +1,141 @@
+(def ^:private bad-currencies
+  "Currencies known not to be supported by the Java [[Currency]] classes. Rendered as USD, then the symbols are
+  replaced."
+  #{:BTC})
```
Perhaps? They may add BTC in later versions, of course. On the other hand we'll need this extra logic until all JVMs we support have all the currencies we do, which sounds like forever. I'd rather keep this as is and avoid possible future JVM version shear.

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers.clj:
```diff
@@ -0,0 +1,141 @@
+(defn- apply-currency-style [text ^Currency currency ^Locale locale style currency-key]
+  (let [sym      (if (= currency-key :BTC) "₿" (.getSymbol currency locale))
+        label    (if (= currency-key :BTC) "Bitcoin" (.getDisplayName currency locale))
+        code     (if (= currency-key :BTC) "BTC" (.getCurrencyCode currency))]
```
Fixed.

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers.clj:
```diff
@@ -0,0 +1,141 @@
+       (subs text (count (.getSymbol (.getCurrency nf) locale)))))
+
+(defn- apply-currency-style [text ^Currency currency ^Locale locale style currency-key]
+  (let [sym      (if (= currency-key :BTC) "₿" (.getSymbol currency locale))
```
Done.

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers.clj:
```diff
@@ -0,0 +1,141 @@
+  [options]
+  (let [currency       (some-> options :currency keyword)
+        locale         (active-locale options)
+        currency-sym   (when currency (symbol-for currency locale))
```
Done.

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers.clj:
```diff
@@ -0,0 +1,141 @@
+        (let [^DecimalFormat df nf ;; Scientific mode always uses the DecimalFormat subclass.
+              exp (str/last-index-of formatted (.getExponentSeparator (.getDecimalFormatSymbols df)))]
+          {:mantissa (subs formatted 0 exp)
+           :exponent (subs formatted (inc exp))})))))
```
Good call.

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers.cljs:
```diff
@@ -0,0 +1,112 @@
+                  :currencyDisplay (:currency-style options)
+                  ;; Always use grouping separators, but we may remove them per number_separators.
+                  :useGrouping              true
+                  :minimumIntegerDigits     (:minimum-integer-digits     options)
+                  :minimumFractionDigits    (:minimum-fraction-digits options default-fraction-digits)
+                  :maximumFractionDigits    (:maximum-fraction-digits options default-fraction-digits)
+                  :minimumSignificantDigits (:minimum-significant-digits options)
+                  :maximumSignificantDigits (:maximum-significant-digits options)})))))
```
Fixed.

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers.clj:
```diff
@@ -0,0 +1,141 @@
+(defn format-number-scientific
+  "Formats a number in scientific notation. The wrangling required differs by platform."
+  [number options]
+  (let [nf   (number-formatter-for-options options)
```
Yes, done.

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers_core.cljc:
```diff
@@ -0,0 +1,31 @@
+  (format-number-basic [this number] "Returns a String that represents the number in this format.")
+  (split-exponent [this formatted]
+                  "Given a scientific notation string, split it at the locale-dependent exponent.
+                  Returns a simple case.")
```
I'm not even sure what I meant by that anymore. It actually returns a map with `:mantissa` and `:exponent` strings. Fixed.

bshepherdson commented on shared/src/metabase/shared/formatting/numbers.cljc:
```diff
@@ -0,0 +1,106 @@
+  (format-number-compact-basic number))
+
+(defmethod format-number-compact "percent" [number _options]
+  (str (format-number-compact-basic (* 100 number)) "%"))
```
Agreed, and yes this is preserving the existing behavior.

bshepherdson commented on shared/src/metabase/shared/formatting/numbers.cljc:
```diff
@@ -0,0 +1,106 @@
+  (let [abs-value (abs number)
+        [power suffix] (first (filter #(>= abs-value (first %)) humanized-powers))]
+    (str (format-number (/ number power) {:minimum-fraction-digits 1 :maximum-fraction-digits 1})
+         suffix)))
```
That makes sense. I've turned this into let-bound inner function.

bshepherdson commented on shared/src/metabase/shared/formatting/internal/numbers.cljs:
```diff
@@ -0,0 +1,112 @@
+        ;; Intl.NumberFormat.formatToParts(1) returns, eg. [currency, integer, decimal, fraction]
+        ;; Drop decimal and fraction, and replace integer's :value with our provided text.
+        (apply str (for [{:keys [type value]} (js->clj (.formatToParts nf 1) :keywordize-keys true)
+                         :when (not (#{"decimal" "fraction"} type))]
```
Ah, this seems better, in case there are other values returned (Exponents? Negatives?)

kulyk commented on frontend/src/metabase/components/FieldValuesWidget.jsx:
```diff
@@ -637,7 +637,7 @@ function renderValue(fields, formatOptions, value, options) {
       value={value}
       column={fields[0]}
-      maximumFractionDigits={20}
+      maximum_fraction_digits={20}
```
Is there anything we can do to keep using camel-case on the frontend as much as possible?
Very few people would expect a React component accepting props in snake case

bshepherdson replied:
Hm. The React props are passed straight through to `formatValue` right now. Since the JS wrappers just forward the options, and all the Clojure(Script) logic is behind `metabase.shared.util/normalize-map`, I could expand that to turn both `"snake_case"` and `"camelCase"` into `:kebab-case` and it should Just Work.

More broadly, the options objects for formatting started out as a mix a camelCase and snake_case, with no particular pattern. (I suspect JSON convention in `visualization_settings` fields leads to snake_case, and React props convention to camelCase.) This porting effort has started out standardizing on snake_case but I think it could be swung the other way if we wanted. @ranquild ?

camsaul replied:
I think `camelCase` makes sense sense TBH. Let's just pick something and stick with it. @bshepherdson, we could use https://github.com/clj-commons/camel-snake-kebab to handle the `camelCase` => `kebab-case` and back transformations

bshepherdson replied:
Okay, I'm using `camel-snake-kebab` as suggested. Now the Clojure code will see uniform `:kebab-case-keywords` whatever the FE code passes as the keys of `options`. I've switched the React prop uses back to camelCase.

The only place that's a bit wonky is in the type for the options, but most of the call sites aren't TS right now, so we can sidestep the problem for now.

bshepherdson committed:
review comments round two

metamben commented on shared/src/metabase/shared/formatting/internal/numbers.clj:
```diff
@@ -0,0 +1,144 @@
+  Returns a [[core/NumberFormatter]] instance for each set of options.
+  These formatters are reusable, but this does no caching."
+  [options]
+  (let [currency       (some-> options :currency keyword)
+        locale         (active-locale options)
+        currency-sym   (some-> currency (symbol-for locale))
+        nf             (number-formatter-for-options-baseline options locale)
+        pre            (preformat-step options)]
```
There seem to be a few redundant spaces here.

metamben commented on shared/src/metabase/shared/util.cljc:
```diff
@@ -12,3 +15,35 @@
+
+  Returns an empty map if nil is input (like [[update-keys]]).
+
+  Note that namespaces on keys are dropped (see [[normalize-key]]). Since this is intended for accepting JSON options
```
They are not quite dropped now.

bshepherdson replied:
Fixed.

metamben commented on shared/src/metabase/shared/util.cljc:
```diff
@@ -12,3 +15,35 @@
+
+  If the input is a namespaced keyword, the namespace is untouched but the base name is converted."
+  [k]
+  (keyword (when (keyword? k) (namespace k))
```
Why not normalize the namespace too? Does that seem surprising or potentially dangerous to you? (I understand that in the currently envisioned cases this won't happen anyway.)

bshepherdson replied:
I'm kind of assuming that if we're getting passed Clojure data structures they'll be named in Clojure style as `:some.name-space/foo`.

metamben reviewed (approved):
Nice work! 👍

bshepherdson committed:
review comments, use csk

bshepherdson committed:
test fix

kulyk commented on frontend/test/metabase/scenarios/question/settings.cy.spec.js:
```diff
@@ -231,8 +231,8 @@ describe("scenarios > question > settings", () => {
 
-      cy.findByText("₿ 2.07");
-      cy.findByText("₿ 6.10");
+      cy.findByText("₿2.07");
+      cy.findByText("₿6.10");
```
Did we change the formatting here on purpose?

kulyk commented on frontend/test/metabase/lib/formatting.unit.spec.js:
```diff
@@ -69,15 +70,50 @@ describe("formatting", () => {
+        type: "cell",
+      };
+
+      it("from positive USD", () => {
+        expect(formatNumber(1234.56, options)).toBe("1,234.56");
+      });
+      it("from negative USD", () => {
+        expect(formatNumber(-1234.56, options)).toBe("-1,234.56");
+      });
```
Please add newlines between tests

bshepherdson replied:
Done.

kulyk commented on frontend/src/metabase/components/SingleSelectListField/SingleSelectListField.unit.spec.js:
```diff
@@ -20,7 +20,7 @@ function renderValue(fields, formatOptions, value, options) {
       value={value}
       column={fields[0]}
-      maximumFractionDigits={20}
+      maximum_fraction_digits={20}
```
We seem to still use the snake case here?

bshepherdson replied:
Ah, good catch. I think I searched for uses of `_significant_digits` but not `_fraction_digits`.
Fixed.

kulyk commented on frontend/src/metabase/visualizations/visualizations/PieChart/PieChart.jsx:
```diff
@@ -371,7 +371,8 @@ export default class PieChart extends Component {
         jsx: true,
         majorWidth: 0,
         number_style: "percent",
-        decimals,
+        minimum_fraction_digits: decimals,
+        maximum_fraction_digits: decimals,
```
Are we still expecting snake cased named options here?

bshepherdson replied:
We can use either, since they get normalized, so we may as well use camelCase.

bshepherdson committed:
review comments

kulyk commented on frontend/test/metabase/lib/formatting.unit.spec.js:
```diff
@@ -1,4 +1,5 @@
-import { isElementOfType } from "react-dom/test-utils";
+import { render, unmountComponentAtNode } from "react-dom";
+import { act, isElementOfType } from "react-dom/test-utils";
```
Can we please use `@testing-library/react` here? It should also let us avoid wrapping every render in `act`

bshepherdson committed:
review: use more ergonomic React test harness

kulyk reviewed (approved):
LGTM, awesome work! 👏

camsaul commented:
Resolves #27180
